hubblestack/status.py

```diff
@@ -0,0 +1,271 @@
+            data.
+
+            The output (after formatting as json) looks like the following
+            (which was truncated and reordered slightly for presentation here).
+
+            .. code-block:: javascript
+                {
+                  …
+                  "hubblestack.daemon.schedule": {
+                    "count": 186, "last_t": 1541773420.481246,
+                    "dt": 0.2783069610595703, "ema_dt": 0.5015859371455758,
+                    "dur": 0.00010395050048828125, "ema_dur": 0.0003155270629760326
+                  },
+                  …
+                  "HEALTH": {
+                    "alive": "yes",
+                    "last_activity": {
+                      "time": 1541773420.481246, "dt": 0.2783069610595703
+                    }
+                  },
+                  "__doc__": {
+                    "service.name.here": {
+                      "count": "number of times the counter was called",
+                      …
+                    },
+                    "HEALTH": {
+                      "last_activity": {
+                        "dt": "the minimum dt across all tracked counters",
+                        …
+                      },
+                      "alive": {
+                        "yes": "something was called within the last 60s",
+                        …
+                      }
+                    }
+                  }
+                }
+        '''
+        r = { x: cls.dat[x].asdict for x in cls.dat }
+        min_dt = min([ x['dt'] for x in r.values() ])
+        max_t  = max([ x['last_t'] for x in r.values() ])
+        h1 = {'time': max_t, 'dt': min_dt}
+        r['HEALTH'] = h2 = {'last_activity': h1}
+        r['__doc__'] = {
+            'service.name.here': {
+                "count": 'number of times the counter was called',
+                "ema_dur": 'average duration of the calls',
+                "dt": 'time since the last call of the counter',
+                "ema_dt": 'average time between calls',
+                "dur": 'duration of the last call',
+                "last_t": 'the last time the counter was called',
+            },
+            'HEALTH': {
+                "last_activity": {
+                    "dt": 'the minimum dt across all tracked counters',
+                    "time": 'the time of the most recent counter',
+                },
+                "alive": {
+                        'yes': 'something was called within the last 60s',
+                        'warn': 'something was called within the last 300s',
+                        'hung': 'nothing has been called in 600s minutes',
+                        'unknown': 'unknown — probably not a good sign though',
+                },
+            },
+        }
+        h2['alive'] = 'unknown'
+        if h1['dt'] < 300:
+            h2['alive'] = 'warn'
+        if h1['dt'] >= 600:
+            h2['alive'] = 'hung'
+        if h1['dt'] < 60:
+            h2['alive'] = 'yes'
```

Comment:
Yeah, these should be configurable for sure. I meant to actually ask what the defaults should be. I picked those numbers out of thin air. I just flaked out on asking the question.

Hubble runs as root. It's definitely possible to print the status to stdout on any user's console; the problem is, afaik, a signal handler doesn't actually get any information about which process issued the signal. If it did, we could trace back to the user shell and print to their pts.

Reply:
Oh ... surely the schedule() runs ever second or so? I marked schedule() so something would be getting updated at least that often unless something set the whole process to iowait...

Reply:
Ah, I missed that you had wrapped it around `schedule()`. We probably shouldn't do that, we don't need to be writing to disk every half second.

Reply:
Oh, and yeah let's log that we wrote the status file. Trying to track down the signaling process is probably unnecessary.

Reply:
it doesn't write to disk every half second... it only writes to disk when it gets the sig10. It simply increments the counters and things when it runs through the mark. Possibly even that's too much? but it seemed trivial to me.

Reply:
Yikes, my brain apparently saw what it wanted to see in a bunch of this code. Sorry for my repeated mistakes in reading the code. Incrementing a counter in memory is fine.

I wonder if there's any situation where hubble will be stuck enough that it won't even respond to the SIGUSR1? If that's a possibility, it might be worth writing the data to disk on a cadence (hourly or something) just so we can see when it was last written and what it looks like. My signal knowledge is not strong enough to know if that's a possibility though.

Reply:
I suppose we could set a timer for that. I can think of reasons to do it, but I don't think it's necessary in this context.

I'm pretty sure the signal handlers fire even if the process is completely stuck. The kernel says, "hey signal" and interrupts the process. The handler fires outside the process execution stack. I've even seen people use signals for multi-process-like environments (http://poe.perl.org/) — though, that's a little janky no matter how well it's done, and POE is done well.

You're limited in what you can do during the interrupt — and I don't actually know those limits — but I'm confident the process can't be too stuck to fire the sigusr1 handler unless the system is unhealthy in some way that affects the kernel operation (disk-io issues and the like).

Reply:
Perfect, thanks.